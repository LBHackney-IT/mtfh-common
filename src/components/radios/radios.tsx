import React, {
  Children,
  ComponentPropsWithoutRef,
  ReactNode,
  cloneElement,
  forwardRef,
  isValidElement,
  useEffect,
  useMemo,
  useRef,
} from 'react';
import cn from 'classnames';
import { Radios as RadiosJs } from 'lbh-frontend';
import mergeRefs from 'react-merge-refs';

import './styles.scss';

export interface RadioProps extends ComponentPropsWithoutRef<'input'> {
  id: string;
  hint?: string;
  children: ReactNode;
  conditionalId?: string;
}

export const Radio = forwardRef<HTMLInputElement, RadioProps>(function Input(
  { id, className, type = 'radio', hint, children, conditionalId, ...props },
  ref
) {
  return (
    <div className={cn('govuk-radios__item', className)}>
      <input
        ref={ref}
        id={id}
        className="govuk-radios__input"
        type={type}
        aria-describedby={hint ? `${id}-hint` : undefined}
        data-aria-controls={conditionalId}
        {...props}
      />
      <label className="govuk-label govuk-radios__label" htmlFor={id}>
        {children}
      </label>
      {hint ? (
        <span id={`${id}-hint`} className="govuk-hint govuk-radios__hint">
          {hint}
        </span>
      ) : null}
    </div>
  );
});

export const RadioDivider = forwardRef<
  HTMLDivElement,
  ComponentPropsWithoutRef<'div'>
>(function RadioDivider(props, ref) {
  return <div ref={ref} className="govuk-radios__divider" {...props} />;
});

export const RadioConditional = forwardRef<
  HTMLDivElement,
  ComponentPropsWithoutRef<'div'>
>(function RadioConditional(props, ref) {
  return (
    <div
      ref={ref}
      className="govuk-radios__conditional govuk-radios__conditional--hidden"
      {...props}
    />
  );
});

export interface RadiosProps extends ComponentPropsWithoutRef<'div'> {
  variant?: 'base' | 'small';
  inline?: boolean;
  name?: string;
  error?: string;
}

export const Radios = forwardRef<HTMLDivElement, RadiosProps>(function Radios(
  { variant = 'base', inline = false, name, children, error, ...props },
  ref
) {
  const localRef = useRef<HTMLDivElement>();

  useEffect(() => {
    if (localRef.current) {
      new RadiosJs(localRef.current).init();
    }
  }, []);

  const hasConditionals = useMemo(
    () =>
      Children.toArray(children).some(
        (child) => isValidElement(child) && child.type === RadioConditional
      ),
    [children]
  );

  return (
    <div
      ref={mergeRefs([localRef, ref])}
      className={cn(
        'govuk-radios',
        {
          'govuk-radios-small': variant === 'small',
          'govuk-radios--inline': inline,
          'govuk-radios--conditionals': hasConditionals,
        },
        'lbh-radios'
      )}
      {...props}
    >
      {Children.map(
        children,
        (child) =>
          isValidElement(child) &&
          cloneElement(child, {
            name,
          })
      )}
    </div>
  );
});
