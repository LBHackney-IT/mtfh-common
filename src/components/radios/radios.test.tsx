import React from 'react';
import { render } from '@testing-library/react';

import { Radio, RadioConditional, RadioDivider, Radios } from './radios';

test('it renders a radio', () => {
  const { container } = render(<Radio id="test">Label</Radio>);
  expect(container).toMatchSnapshot();
});

test('it renders a radio with a hint', () => {
  const { container } = render(
    <Radio id="test" hint="A short description">
      Label
    </Radio>
  );
  expect(container).toMatchSnapshot();
});

test('it renders a group of radios', () => {
  const { container } = render(
    <Radios>
      <Radio id="test">Label</Radio>
      <Radio id="test2">Label2</Radio>
    </Radios>
  );
  expect(container).toMatchSnapshot();
});

test('it renders a group of radios with a divider', () => {
  const { container } = render(
    <Radios>
      <Radio id="test">Label</Radio>
      <RadioDivider>Or</RadioDivider>
      <Radio id="test2">Label2</Radio>
    </Radios>
  );
  expect(container).toMatchSnapshot();
});

test('it renders a group of radios with a conditional', () => {
  const { container } = render(
    <Radios>
      <Radio id="test" conditionalId="conditional">
        Label
      </Radio>
      <RadioConditional id="conditional">Hello</RadioConditional>
      <Radio id="test2">Label2</Radio>
    </Radios>
  );
  expect(container).toMatchSnapshot();
});
