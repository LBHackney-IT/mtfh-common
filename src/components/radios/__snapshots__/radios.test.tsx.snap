// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders a group of radios 1`] = `
<div>
  <div
    class="govuk-radios lbh-radios"
  >
    <div
      class="govuk-radios__item"
    >
      <input
        class="govuk-radios__input"
        id="test"
        type="radio"
      />
      <label
        class="govuk-label govuk-radios__label"
        for="test"
      >
        Label
      </label>
    </div>
    <div
      class="govuk-radios__item"
    >
      <input
        class="govuk-radios__input"
        id="test2"
        type="radio"
      />
      <label
        class="govuk-label govuk-radios__label"
        for="test2"
      >
        Label2
      </label>
    </div>
  </div>
</div>
`;

exports[`it renders a group of radios with a conditional 1`] = `
<div>
  <div
    class="govuk-radios govuk-radios--conditionals lbh-radios"
  >
    <div
      class="govuk-radios__item"
    >
      <input
        aria-controls="conditional"
        aria-expanded="false"
        class="govuk-radios__input"
        id="test"
        type="radio"
      />
      <label
        class="govuk-label govuk-radios__label"
        for="test"
      >
        Label
      </label>
    </div>
    <div
      class="govuk-radios__conditional govuk-radios__conditional--hidden"
      id="conditional"
    >
      Hello
    </div>
    <div
      class="govuk-radios__item"
    >
      <input
        class="govuk-radios__input"
        id="test2"
        type="radio"
      />
      <label
        class="govuk-label govuk-radios__label"
        for="test2"
      >
        Label2
      </label>
    </div>
  </div>
</div>
`;

exports[`it renders a group of radios with a divider 1`] = `
<div>
  <div
    class="govuk-radios lbh-radios"
  >
    <div
      class="govuk-radios__item"
    >
      <input
        class="govuk-radios__input"
        id="test"
        type="radio"
      />
      <label
        class="govuk-label govuk-radios__label"
        for="test"
      >
        Label
      </label>
    </div>
    <div
      class="govuk-radios__divider"
    >
      Or
    </div>
    <div
      class="govuk-radios__item"
    >
      <input
        class="govuk-radios__input"
        id="test2"
        type="radio"
      />
      <label
        class="govuk-label govuk-radios__label"
        for="test2"
      >
        Label2
      </label>
    </div>
  </div>
</div>
`;

exports[`it renders a radio 1`] = `
<div>
  <div
    class="govuk-radios__item"
  >
    <input
      class="govuk-radios__input"
      id="test"
      type="radio"
    />
    <label
      class="govuk-label govuk-radios__label"
      for="test"
    >
      Label
    </label>
  </div>
</div>
`;

exports[`it renders a radio with a hint 1`] = `
<div>
  <div
    class="govuk-radios__item"
  >
    <input
      aria-describedby="test-hint"
      class="govuk-radios__input"
      id="test"
      type="radio"
    />
    <label
      class="govuk-label govuk-radios__label"
      for="test"
    >
      Label
    </label>
    <span
      class="govuk-hint govuk-radios__hint"
      id="test-hint"
    >
      A short description
    </span>
  </div>
</div>
`;
